WalletWasabi/WabiSabi/Client/ArenaClient.cs

```diff
@@ -94,5 +94,43 @@ public async Task ConfirmConnectionAsync(Guid roundId, Guid aliceId, IEnumerable
 			AmountCredentialClient.HandleResponse(confirmConnectionResponse.ZeroAmountCredentials, zeroAmountCredentialResponseValidation);
 			WeightCredentialClient.HandleResponse(confirmConnectionResponse.ZeroWeightCredentials, zeroWeightCredentialResponseValidation);
 		}
+
+		public async Task SignTransactionAsync(Guid roundId, IEnumerable<ICoin> coinsToSign, BitcoinSecret bitcoinSecret, Transaction unsignedCoinJoin)
+		{
+			if (unsignedCoinJoin.Inputs.Count == 0)
+			{
+				throw new ArgumentException("No inputs to sign.", nameof(unsignedCoinJoin));
+			}
+
+			if (!coinsToSign.Any())
+			{
+				throw new ArgumentException("No coins were provided.", nameof(coinsToSign));
+			}
+
+			var myInputs = coinsToSign.ToDictionary(c => c.Outpoint);
+			var signedCoinJoin = unsignedCoinJoin.Clone();
+			List<InputWitnessPair> signatures = new();
+
+			foreach (var txInput in signedCoinJoin.Inputs.AsIndexedInputs().Where(input => myInputs.ContainsKey(input.PrevOut)))
+			{
+				var coin = myInputs[txInput.PrevOut];
+
+				signedCoinJoin.Sign(bitcoinSecret, coin);
+
+				if (!txInput.VerifyScript(coin, out _))
+				{
+					throw new InvalidOperationException("Witness is missing.");
+				}
+
+				signatures.Add(new InputWitnessPair(txInput.Index, txInput.WitScript));
+			}
+
+			if (myInputs.Count != signatures.Count)
+			{
+				throw new InvalidOperationException("Missing signatures.");
+			}
```

Comment:
This validation makes sense but we can do it before signing the inputs because we already know if this is true or false early, after all `signatures.Count` is equal to `signedCoinJoin.Inputs.AsIndexedInputs().Where(input => myInputs.ContainsKey(input.PrevOut)).Count()` so, we can save the `IndexedInputs` array in a variable and check the counts match and then we can use the same variable for the foreach loop.

Reply:
agreed, better to validate as much as possible before making any signatures

Reply:
ACK!